types/int_test.go

```diff
@@ -109,3 +110,342 @@ func TestUint(t *testing.T) {
+	if i1 < i2 {
+		return i1
+	}
+	return i2
+}
+
+func TestArithUint(t *testing.T) {
+	for d := 0; d < 20; d++ {
+		n1 := uint64(rand.Uint32())
+		i1 := NewUint(n1)
+		n2 := uint64(rand.Uint32())
+		i2 := NewUint(n2)
+
+		cases := []struct {
+			ires Uint
+			nres uint64
+		}{
+			{i1.Add(i2), n1 + n2},
+			{i1.Mul(i2), n1 * n2},
+			{i1.Div(i2), n1 / n2},
+			{i1.AddRaw(n2), n1 + n2},
+			{i1.MulRaw(n2), n1 * n2},
+			{i1.DivRaw(n2), n1 / n2},
+			{MinUint(i1, i2), minuint(n1, n2)},
+		}
+
+		for _, tc := range cases {
+			require.Equal(t, tc.nres, tc.ires.Uint64())
+		}
+
+		if n2 > n1 {
+			continue
+		}
+
+		subs := []struct {
+			ires Uint
+			nres uint64
+		}{
+			{i1.Sub(i2), n1 - n2},
+			{i1.SubRaw(n2), n1 - n2},
+		}
+
+		for _, tc := range subs {
+			require.Equal(t, tc.nres, tc.ires.Uint64())
+		}
+	}
+}
+
+func TestCompUint(t *testing.T) {
+	for d := 0; d < 20; d++ {
+		n1 := rand.Uint64()
+		i1 := NewUint(n1)
+		n2 := rand.Uint64()
+		i2 := NewUint(n2)
+
+		cases := []struct {
+			ires bool
+			nres bool
+		}{
+			{i1.Equal(i2), n1 == n2},
+			{i1.GT(i2), n1 > n2},
+			{i1.LT(i2), n1 < n2},
+		}
+
+		for _, tc := range cases {
+			require.Equal(t, tc.nres, tc.ires)
+		}
+	}
+}
+
+func randint() Int {
+	return NewInt(rand.Int63())
+}
+
+func TestImmutabilityInt(t *testing.T) {
+	ops := []func(*Int){
+		func(i *Int) { _ = i.Add(randint()) },
+		func(i *Int) { _ = i.Sub(randint()) },
+		func(i *Int) { _ = i.Mul(randint()) },
+		func(i *Int) { _ = i.Div(randint()) },
+		func(i *Int) { _ = i.AddRaw(rand.Int63()) },
+		func(i *Int) { _ = i.SubRaw(rand.Int63()) },
+		func(i *Int) { _ = i.MulRaw(rand.Int63()) },
+		func(i *Int) { _ = i.DivRaw(rand.Int63()) },
+		func(i *Int) { _ = i.Neg() },
+		func(i *Int) { _ = i.IsZero() },
+		func(i *Int) { _ = i.Sign() },
+		func(i *Int) { _ = i.Equal(randint()) },
+		func(i *Int) { _ = i.GT(randint()) },
+		func(i *Int) { _ = i.LT(randint()) },
+		func(i *Int) { _ = i.String() },
+	}
+
+	for i := 0; i < 20; i++ {
+		n := rand.Int63()
+		ni := NewInt(n)
+
+		for _, op := range ops {
+			op(&ni)
+
+			require.Equal(t, n, ni.Int64())
+			require.Equal(t, NewInt(n), ni)
+		}
+	}
+}
+
+func TestImmutabilityUint(t *testing.T) {
+	ops := []func(*Uint){
+		func(i *Uint) { _ = i.Add(randuint()) },
+		func(i *Uint) { _ = i.Sub(NewUint(rand.Uint64() % i.Uint64())) },
+		func(i *Uint) { _ = i.Mul(randuint()) },
+		func(i *Uint) { _ = i.Div(randuint()) },
+		func(i *Uint) { _ = i.AddRaw(rand.Uint64()) },
+		func(i *Uint) { _ = i.SubRaw(rand.Uint64() % i.Uint64()) },
+		func(i *Uint) { _ = i.MulRaw(rand.Uint64()) },
+		func(i *Uint) { _ = i.DivRaw(rand.Uint64()) },
+		func(i *Uint) { _ = i.IsZero() },
+		func(i *Uint) { _ = i.Sign() },
+		func(i *Uint) { _ = i.Equal(randuint()) },
+		func(i *Uint) { _ = i.GT(randuint()) },
+		func(i *Uint) { _ = i.LT(randuint()) },
+		func(i *Uint) { _ = i.String() },
+	}
+
+	for i := 0; i < 20; i++ {
+		n := rand.Uint64()
+		ni := NewUint(n)
+
+		for _, op := range ops {
```

Comment:
Nice job! However to fully check immutability here, we need to add a bit more here to get it fully right. Add a parameter to the operations so that it can perform the same operation on our known reference big.Rat. (As that is what justifies any of this randomized testing)

Then we need to maintain a list of all the previous `ni` values, and ensure that each `ni` value equals the corresponding calculated version on the known-reference implementation. We need this because we need to ensure an operation doesn't mutate any previous operation. While this is unlikely, I think its quite important.

Then we may want to consider taking a random walk of length n through the possible operations. However this is a long integration test, and is probably overkill for something thats probably right.